sw/cheri/tests/pinmux_tests.hh

```diff
@@ -0,0 +1,546 @@
+  constexpr uint8_t PmxArduinoGpio0ToAhTmpio0 = 1;
+
+  constexpr GpioPin GpioPinInput  = {2, 0};
+  constexpr GpioPin GpioPinOutput = {2, 1};
+
+  int failures = 0;
+
+  // Set the Arduino GPIO D0 as input and D1 as output.
+  set_gpio_output_enable(gpio, GpioPinOutput, true);
+  set_gpio_output_enable(gpio, GpioPinInput, false);
+
+  // Mux UART3 over Arduino Shield D0 (RX) & D1 (TX)
+  if (!pinmux->block_input_select(SonataPinmux::BlockInput::UartReceive3, PmxUartReceive3ToAhTmpio0)) failures++;
+  if (!pinmux->output_pin_select(SonataPinmux::OutputPin::ArduinoShieldD1, PmxArduinoD1ToUartTx3)) failures++;
+
+  // Test that UART3 works over the muxed Arduino Shield D0 & D1 pins,
+  // and that GPIO does not work, as these pins are not muxed for GPIO.
+  if (!uart_send_receive_test(prng, uart3, UartTimeoutMsec, UartTestBytes)) failures++;
+  if (gpio_write_read_test(gpio, GpioPinOutput, GpioPinInput, GpioWaitMsec, GpioTestLength)) failures++;
+
+  // Mux GPIO over Arduino Shield D0 (GPIO input) & D1 (GPIO output)
+  if (!pinmux->block_input_select(SonataPinmux::BlockInput::ArduinoShieldGpio0, PmxArduinoGpio0ToAhTmpio0)) failures++;
+  if (!pinmux->output_pin_select(SonataPinmux::OutputPin::ArduinoShieldD1, PmxArduinoD1ToGpio_1_1)) failures++;
+
+  // Test that UART3 no longer works (no longer muxed over D0 & D1),
+  // and that our muxed GPIO now works.
+  if (uart_send_receive_test(prng, uart3, UartTimeoutMsec, UartTestBytes)) failures++;
+  if (!gpio_write_read_test(gpio, GpioPinOutput, GpioPinInput, GpioWaitMsec, GpioTestLength)) failures++;
+
+  // Mux back to UART3 again, and thest that UART again passes and GPIO fails.
+  if (!pinmux->block_input_select(SonataPinmux::BlockInput::UartReceive3, PmxUartReceive3ToAhTmpio0)) failures++;
+  if (!pinmux->output_pin_select(SonataPinmux::OutputPin::ArduinoShieldD1, PmxArduinoD1ToUartTx3)) failures++;
+  if (!uart_send_receive_test(prng, uart3, UartTimeoutMsec, UartTestBytes)) failures++;
+  if (gpio_write_read_test(gpio, GpioPinOutput, GpioPinInput, GpioWaitMsec, GpioTestLength)) failures++;
+
+  return failures;
+}
+
+/**
+ * Run the whole suite of pinmux tests.
+ */
+void pinmux_tests(CapRoot root, Log &log) {
+  // Create a bounded capability for pinmux & initialise the driver
+  Capability<volatile uint8_t> pinmux = root.cast<volatile uint8_t>();
+  pinmux.address()                    = PINMUX_ADDRESS;
+  pinmux.bounds()                     = PINMUX_BOUNDS;
+  SonataPinmux Pinmux                 = SonataPinmux(pinmux);
+
+  // Create bounded capabilities for other devices, to be used in testing.
+  Capability<volatile SonataSpi> spi = root.cast<volatile SonataSpi>();
+  spi.address()                      = SPI_ADDRESS;
+  spi.bounds()                       = SPI_BOUNDS;
+
+  Capability<volatile SonataGpioFull> gpio_full = root.cast<volatile SonataGpioFull>();
+  gpio_full.address()                           = GPIO_ADDRESS;
+  gpio_full.bounds()                            = GPIO_FULL_BOUNDS;
```

Comment:
For `SonataSpi` I agree, for `SonataGpioFull` this is essentially a small temporary version of the driver sitting in a PR at https://github.com/lowRISC/cheriot-rtos/pull/13, so I think when that is merged and we bump the revision of CHERIoT RTOS used in `sonata-system` we should update all CHERI GPIO usages (including this) to go through that.

Reply:
From checking the code now it appears that I kept `SonataSpi` here because the `SpiFlash` util constructor expects a capability `Capability<volatile SonataSpi>` to the SPI, whereas the `spi_ptr` function only returns a pointer. I'm not sure if there is an easy way to make that conversion - please let me know if there is and I'll happily change this.